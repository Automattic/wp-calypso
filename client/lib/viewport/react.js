/** @format */
/**
 * External dependencies
 */
import React, { forwardRef, useState, useEffect } from 'react';
import { createHigherOrderComponent } from '@wordpress/compose';

/**
 * Internal dependencies
 */
import {
	isWithinBreakpoint,
	subscribeIsWithinBreakpoint,
	MOBILE_BREAKPOINT,
	DESKTOP_BREAKPOINT,
} from '.';

/**
 * React hook for getting the status for a breakpoint and keeping it updated.
 *
 * @param {String} breakpoint The breakpoint to consider.
 *
 * @returns {Boolean} The current status for the breakpoint.
 */
export function useBreakpoint( breakpoint ) {
	const [ state, setState ] = useState( () => ( {
		isActive: isWithinBreakpoint( breakpoint ),
		breakpoint,
	} ) );

	useEffect( () => {
		function handleBreakpointChange( isActive ) {
			setState( prevState => {
				// Ensure we bail out without rendering if nothing changes, by preserving state.
				if ( prevState.isActive === isActive && prevState.breakpoint === breakpoint ) {
					return prevState;
				}
				return { isActive, breakpoint };
			} );
		}

		const unsubscribe = subscribeIsWithinBreakpoint( breakpoint, handleBreakpointChange );
		// The unsubscribe function is the entire cleanup for the effect.
		return unsubscribe;
	}, [ breakpoint ] );

	return breakpoint === state.breakpoint ? state.isActive : isWithinBreakpoint( breakpoint );
}

/**
 * React hook for getting the status for the mobile breakpoint and keeping it
 * updated.
 *
 * @returns {Boolean} The current status for the breakpoint.
 */
export function useMobileBreakpoint() {
	return useBreakpoint( MOBILE_BREAKPOINT );
}

/**
 * React hook for getting the status for the desktop breakpoint and keeping it
 * updated.
 *
 * @returns {Boolean} The current status for the breakpoint.
 */
export function useDesktopBreakpoint() {
	return useBreakpoint( DESKTOP_BREAKPOINT );
}

/**
 * React higher order component for getting the status for a breakpoint and
 * keeping it updated.
 *
 * @param {String} breakpoint The breakpoint to consider.
 *
 * @returns {Function} A function that given a component returns the
 * wrapped component.
 */
export const withBreakpoint = breakpoint =>
	createHigherOrderComponent(
		WrappedComponent =>
			forwardRef( ( props, ref ) => {
				const isActive = useBreakpoint( breakpoint );
				return <WrappedComponent { ...props } isBreakpointActive={ isActive } ref={ ref } />;
			} ),
		'WithBreakpoint'
	);

/**
 * React higher order component for getting the status for the mobile
 * breakpoint and keeping it updated.
 *
 * @param {React.Component|Function} Wrapped The component to wrap.
 *
 * @returns {Function} The wrapped component.
 */
export const withMobileBreakpoint = createHigherOrderComponent(
	WrappedComponent =>
		forwardRef( ( props, ref ) => {
			const isActive = useBreakpoint( MOBILE_BREAKPOINT );
			return <WrappedComponent { ...props } isBreakpointActive={ isActive } ref={ ref } />;
		} ),
	'WithMobileBreakpoint'
);

/**
 * React higher order component for getting the status for the desktop
 * breakpoint and keeping it updated.
 *
 * @param {React.Component|Function} Wrapped The component to wrap.
 *
 * @returns {Function} The wrapped component.
 */
export const withDesktopBreakpoint = createHigherOrderComponent(
	WrappedComponent =>
		forwardRef( ( props, ref ) => {
			const isActive = useBreakpoint( DESKTOP_BREAKPOINT );
			return <WrappedComponent { ...props } isBreakpointActive={ isActive } ref={ ref } />;
		} ),
	'WithDesktopBreakpoint'
);
