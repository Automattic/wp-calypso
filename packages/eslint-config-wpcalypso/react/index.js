/** @format */

module.exports = {
	extends: '../index.js',
	parserOptions: {
		ecmaVersion: 7,
		ecmaFeatures: {
			jsx: true,
		},
		sourceType: 'module',
	},
	plugins: [ 'react', 'react-hooks', 'wpcalypso' ],
	rules: {
		'react/jsx-curly-spacing': [ 2, 'always' ],
		'react/jsx-no-duplicate-props': 2,
		'react/jsx-no-target-blank': 2,
		'react/jsx-no-undef': 2,
		'react/jsx-tag-spacing': 2,
		'react/jsx-uses-react': 2,
		'react/jsx-uses-vars': 2,
		'react/no-danger': 2,
		'react/no-deprecated': 2,
		'react/no-did-mount-set-state': 2,
		'react/no-did-update-set-state': 2,
		'react/no-is-mounted': 2,
		'react/no-string-refs': 2,
		'react/prefer-es6-class': 2,
		'react/react-in-jsx-scope': 2,
		'react-hooks/rules-of-hooks': 2,
		'react-hooks/exhaustive-deps': 1,
	},
};
