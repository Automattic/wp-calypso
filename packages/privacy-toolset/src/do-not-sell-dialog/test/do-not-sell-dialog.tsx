/**
 * @jest-environment jsdom
 */
import { render, fireEvent } from '@testing-library/react';
import { unmountComponentAtNode } from 'react-dom';
import Modal from 'react-modal';
import '@testing-library/jest-dom';
import { DoNotSellDialog } from '..';

const genericContent = {
	title: 'title',
	longDescription: (
		<>
			<p>paragraph 1</p>
			<p>paragraph 2</p>
			<p>paragraph 3</p>
		</>
	),
	toggleLabel: 'toggleLabel',
	closeButton: 'closeButton',
};

describe( 'DoNotSellDialog', () => {
	let modalRoot: HTMLDivElement | null = null;

	beforeEach( () => {
		modalRoot = document.createElement( 'div' );
		modalRoot.setAttribute( 'id', 'modal-root' );
		document.body.appendChild( modalRoot );
		Modal.setAppElement( modalRoot );
	} );

	afterEach( () => {
		if ( null !== modalRoot ) {
			unmountComponentAtNode( modalRoot );
			document.body.removeChild( modalRoot );
			modalRoot = null;
		}
	} );

	test( 'renders correctly', () => {
		const { baseElement, getByText } = render(
			<DoNotSellDialog
				content={ genericContent }
				onToggleActive={ jest.fn() }
				onClose={ jest.fn() }
				isOpen
			/>
		);

		expect( getByText( 'title' ) ).toBeInTheDocument();
		expect( getByText( 'paragraph 1' ) ).toBeInTheDocument();
		expect( getByText( 'paragraph 2' ) ).toBeInTheDocument();
		expect( getByText( 'paragraph 3' ) ).toBeInTheDocument();
		expect( getByText( 'toggleLabel' ) ).toBeInTheDocument();
		expect( getByText( 'closeButton' ) ).toBeInTheDocument();
		expect( baseElement ).toMatchSnapshot();
	} );
	test( 'fires onClose() on text button click', () => {
		const onClose = jest.fn();
		const { getByText } = render(
			<DoNotSellDialog
				content={ genericContent }
				onToggleActive={ jest.fn() }
				onClose={ onClose }
				isOpen
			/>
		);

		fireEvent.click( getByText( 'closeButton' ) );
		expect( onClose ).toHaveBeenCalled();
	} );
	test( 'fires onClose() on "x" button click', () => {
		const onClose = jest.fn();
		const { baseElement: container } = render(
			<DoNotSellDialog
				content={ genericContent }
				onToggleActive={ jest.fn() }
				onClose={ onClose }
				isOpen
			/>
		);

		const closeButton = container.querySelector( '.do-not-sell__close-button' );
		expect( closeButton ).not.toBeNull();

		fireEvent.click( closeButton! );
		expect( onClose ).toHaveBeenCalled();
	} );
	test( 'fires onToggleActive(true) on toggle click when isActive=false', () => {
		const onToggleActive = jest.fn();
		const { getByText } = render(
			<DoNotSellDialog
				content={ genericContent }
				isActive={ false }
				onToggleActive={ onToggleActive }
				onClose={ jest.fn() }
				isOpen
			/>
		);

		fireEvent.click( getByText( 'toggleLabel' ) );
		expect( onToggleActive ).toHaveBeenCalledWith( true );
	} );
	test( 'fires onToggleActive(false) on toggle click when isActive=true', () => {
		const onToggleActive = jest.fn();
		const { getByText } = render(
			<DoNotSellDialog
				content={ genericContent }
				isActive={ true }
				onToggleActive={ onToggleActive }
				onClose={ jest.fn() }
				isOpen
			/>
		);

		fireEvent.click( getByText( 'toggleLabel' ) );
		expect( onToggleActive ).toHaveBeenCalledWith( false );
	} );
} );
