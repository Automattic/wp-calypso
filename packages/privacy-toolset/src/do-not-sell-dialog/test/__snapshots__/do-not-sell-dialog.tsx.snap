// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoNotSellDialog renders correctly 1`] = `
<body
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <p
    class="a11y-speak-intro-text"
    hidden="hidden"
    id="a11y-speak-intro-text"
    style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
  >
    Notifications
  </p>
  <div
    aria-atomic="true"
    aria-live="assertive"
    aria-relevant="additions text"
    class="a11y-speak-region"
    id="a11y-speak-assertive"
    style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
  />
  <div
    aria-atomic="true"
    aria-live="polite"
    aria-relevant="additions text"
    class="a11y-speak-region"
    id="a11y-speak-polite"
    style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
  />
  <div
    aria-hidden="true"
    id="modal-root"
  />
  <div />
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay do-not-sell"
    >
      <div
        aria-describedby="do-not-sell-description"
        aria-labelledby="do-not-sell-title"
        aria-modal="true"
        class="ReactModal__Content do-not-sell__dialog"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="do-not-sell__header"
        >
          <h3
            class="do-not-sell__title"
            id="do-not-sell-title"
          >
            title
          </h3>
          <button
            aria-label="Close dialog"
            class="do-not-sell__close-button"
          />
        </div>
        <div
          class="do-not-sell__content"
          id="do-not-sell-description"
        >
          <p>
            paragraph 1
          </p>
          <p>
            paragraph 2
          </p>
          <p>
            paragraph 3
          </p>
        </div>
        <label
          class="do-not-sell__preference"
        >
          <span
            class="components-form-toggle do-not-sell__checkbox"
          >
            <input
              class="components-form-toggle__input"
              type="checkbox"
            />
            <span
              class="components-form-toggle__track"
            />
            <span
              class="components-form-toggle__thumb"
            />
          </span>
          toggleLabel
        </label>
        <div
          class="do-not-sell__footer"
        >
          <button
            class="do-not-sell__button"
          >
            closeButton
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
